### ISSUE TITLE
Possible router bug since 1.3.0

### ISSUE THREAD

**Issue by atikhono** [Kong/kong#5371]
### Summary

Prefix path matching behaviour has changed since 1.3.0. Longest path routes are no longer matched as expected. Reproducible on 1.4.x as well. 1.2.2 is not affected.

### Steps To Reproduce

1. Start kong in db-less mode with the following kong.conf and kong.yaml:

```
docker run -p 8000:8000 -p 8001:8001 -v `pwd`/kong.yaml:/config/kong.yaml -v `pwd`/kong.conf:/config/kong.conf kong:1.3.0 kong start -c /config/kong.conf
```

kong.conf:

```
database = off
declarative_config = /config/kong.yaml
proxy_access_log = /dev/stdout
admin_listen = 0.0.0.0:8001
admin_access_log = /dev/stdout
```

kong.yaml:

```
_format_version: "1.1"

services:
  - name: devportal-public
    host: localhost
    port: 8001
    path: /routes
    routes:
      - name: public-apiv1
        hosts:
          - api.local
        paths:
          - /rest/devportal/api/v1
          - /rest/devportal

  - name: devportal-aux
    host: localhost
    port: 8001
    path: /services
    routes:
      - name: aux
        hosts:
          - api.local
        paths:
          - /rest/devportal/aux
```
2. Call kong as follows:

```
curl -H 'Host:api.local' localhost:8000/rest/devportal/aux
```

With Kong 1.3.0+ public-apiv1 is getting matched.
With Kong 1.2.2 aux route is getting matched. 

I have found out that matching documentation differs for 1.2.2 and 1.3.0+: https://docs.konghq.com/1.3.x/proxy/#matching-priorities.

> A has only “plain” Host headers and B has has one or more “wildcard” host headers
> A has more non-Host headers than B.
> A has at least one “regex” paths and B has only “plain” paths.
> A’s longer path is longer than B’s longer path.
> A.created_at < B.created_at

Looks like I'm hitting "A’s longer path is longer than B’s longer path"... If I add a regex to aux route path (like so -- /rest/devportal/aux(.*)), it's matched as expected with Kong 1.3.0+. If I add one more rule to aux route (methods: - GET) -- again matched as expected.

Would love to know if this is an intended behaviour change or a bug (look like a bug to me).
If not a bug, please tell me how do I properly fix the declarative configuration piece above to get Kong to match aux route for api.local/rest/devportal/aux?

**Comment by hishamhm** [comment]
> Would love to know if this is an intended behaviour change or a bug (look like a bug to me).
> If not a bug, please tell me how do I properly fix the declarative configuration piece above to get Kong to match aux route for api.local/rest/devportal/aux?

@atikhono Hi, thank you for the report! We need to investigate this a bit further to confirm if it's a bug, but in the meantime, you can get your desired behavior using regex matches. I believe this will work as you intend:

```
_format_version: "1.1"

services:
  - name: devportal-public
    host: localhost
    port: 8001
    path: /routes
    routes:
      - name: public-apiv1
        hosts:
          - api.local
        paths:
          - /rest/devportal/api/v1/?.*$
          - /rest/devportal

  - name: devportal-aux
    host: localhost
    port: 8001
    path: /services
    routes:
      - name: aux
        hosts:
          - api.local
        paths:
          - /rest/devportal/aux/?.*$
```

**Comment by bungle** [comment]
@atikhono mind you trying: #5383

**Comment by atikhono** [comment]
@bungle Sorry for delay, public holidays in my location. I tested my reproducer with a centos-based image and kong-nightly/2020-01-08 build, looks good. I will test it with my application ASAP and let you know in case of any further routing issues. Could we please have this fix included for the upcoming 1.4.x and 2.0.0 releases?
